RPCCore$RPCException: encoded string too long

Just giving 0.6.28 a whirl. Now that underlying RPCCore exceptions are being reported (awesome) we might finally be able to track down the root causes of stuff that's been around for a while. Exciting! First cab off the rank:

```
[error] org.scalajs.testcommon.RPCCore$RPCException: java.io.UTFDataFormatException: encoded string too long: 93549 bytes
[error] 	at org.scalajs.testcommon.RPCCore.$anonfun$handleMessage$4(RPCCore.scala:79)
[error] 	at scala.Option.foreach(Option.scala:257)
[error] 	at org.scalajs.testcommon.RPCCore.$anonfun$handleMessage$1(RPCCore.scala:77)
[error] 	at org.scalajs.testcommon.RPCCore.$anonfun$handleMessage$1$adapted(RPCCore.scala:57)
[error] 	at org.scalajs.testcommon.Serializer$.withInputStream(Serializer.scala:60)
[error] 	at org.scalajs.testcommon.RPCCore.handleMessage(RPCCore.scala:57)
[error] 	at org.scalajs.testadapter.ComJSEnvRPC$$anon$1.run(ComJSEnvRPC.scala:34)
[error] Caused by: java.io.UTFDataFormatException: encoded string too long: 93549 bytes
[error] 	at <jscode>.fillInStackTrace__jl_Throwable(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/webapp-base-test-test-fastopt.js:118762)
[error] 	at <jscode>.init___T__jl_Throwable__Z__Z(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/webapp-base-test-test-fastopt.js:118889)
[error] 	at <jscode>.init___T(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/webapp-base-test-test-fastopt.js:192257)
[error] 	at <jscode>.writeUTF__T__V(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/webapp-base-test-test-fastopt.js:247780)
[error] 	at <jscode>.serialize__O__Ljava_io_DataOutputStream__V(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/webapp-base-test-test-fastopt.js:100981)
[error] 	at <jscode>.makeRPCMsg__p1__B__J__O__Lorg_scalajs_testcommon_Serializer__T(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/webapp-base-test-test-fastopt.js:24948)
[error] 	at <jscode>.file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/webapp-base-test-test-fastopt.js:24912:82(Unknown Source)
[error] 	at <jscode>.apply__O__O(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/webapp-base-test-test-fastopt.js:164688)
[error] 	at <jscode>.map__F1__s_util_Try(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/webapp-base-test-test-fastopt.js:246480)
[error] 	at <jscode>.makeReply__p1__J__s_util_Try__Lorg_scalajs_testcommon_Serializer__T(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/webapp-base-test-test-fastopt.js:24910)
[error] 	at <jscode>.file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/webapp-base-test-test-fastopt.js:24829:109(Unknown Source)
[error] 	at <jscode>.apply__O__O(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/webapp-base-test-test-fastopt.js:164688)
[error] 	at <jscode>.run__V(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/webapp-base-test-test-fastopt.js:163437)
[error] 	at <jscode>.scala$scalajs$concurrent$QueueExecutionContext$PromisesExecutionContext$$$anonfun$execute$2__sr_BoxedUnit__jl_Runnable__sjs_js_$bar(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/webapp-base-test-test-fastopt.js:191307)
[error] 	at <jscode>.file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/webapp-base-test-test-fastopt.js:191324:155(Unknown Source)
[error] 	at <jscode>.h(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/test-classes/webapp-base-test.js:1)
[error] 	at <jscode>.file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/test-classes/webapp-base-test.js:1:225079(Unknown Source)
[error] 	at <jscode>.run(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/test-classes/webapp-base-test.js:1)
[error] 	at <jscode>.d(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.12/test-classes/webapp-base-test.js:1)
```

java.io.DataOutputStream.writeUTF has a limit of 65535 because it first writes a two bit size ([see](https://stackoverflow.com/questions/22741556/dataoutputstream-purpose-of-the-encoded-string-too-long-restriction)) which means the RPC machinery might need some logic to split, send in batches and reconstruct large strings.

Or maybe it could just truncate messages for an easier, quicker, non-crashing solution if it's just std{out,err} piping?

Urgh, I'm running into this @#!$*&#@ problem again. This time I've managed to get a stack trace out! :tada: 

```
[error] org.scalajs.testing.common.RPCCore$RPCException: java.io.UTFDataFormatException: encoded string too long: 65815 bytes
[error] 	at org.scalajs.testing.common.RPCCore.$anonfun$handleMessage$4(RPCCore.scala:78)
[error] 	at scala.Option.foreach(Option.scala:407)
[error] 	at org.scalajs.testing.common.RPCCore.$anonfun$handleMessage$1(RPCCore.scala:76)
[error] 	at org.scalajs.testing.common.RPCCore.$anonfun$handleMessage$1$adapted(RPCCore.scala:56)
[error] 	at org.scalajs.testing.common.Serializer$.withInputStream(Serializer.scala:60)
[error] 	at org.scalajs.testing.common.RPCCore.handleMessage(RPCCore.scala:56)
[error] 	at org.scalajs.testing.adapter.JSEnvRPC.$anonfun$run$1(JSEnvRPC.scala:25)
[error] 	at org.scalajs.testing.adapter.JSEnvRPC.$anonfun$run$1$adapted(JSEnvRPC.scala:25)
[error] 	at org.scalajs.jsenv.nodejs.ComRun$$anon$1.run(ComSupport.scala:74)
[error] Caused by: java.io.UTFDataFormatException: encoded string too long: 65815 bytes
[error] 	at java.io.DataOutputStream.writeUTF(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.13/webapp-base-test-test-fastopt.js:364033)
[error] 	at org.scalajs.testing.common.Serializer$StringSerializer$.serialize(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.13/webapp-base-test-test-fastopt.js:136832)
[error] 	at org.scalajs.testing.common.RPCCore.makeRPCMsg(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.13/webapp-base-test-test-fastopt.js:16291)
[error] 	at <jscode>.{anonymous}()(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.13/webapp-base-test-test-fastopt.js:16268)
[error] 	at scala.scalajs.runtime.AnonFunction1.apply(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.13/webapp-base-test-test-fastopt.js:190147)
[error] 	at scala.util.Success.map(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.13/webapp-base-test-test-fastopt.js:287210)
[error] 	at org.scalajs.testing.common.RPCCore.makeReply(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.13/webapp-base-test-test-fastopt.js:16268)
[error] 	at <jscode>.{anonymous}()(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.13/webapp-base-test-test-fastopt.js:16750)
[error] 	at scala.scalajs.runtime.AnonFunction1.apply(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.13/webapp-base-test-test-fastopt.js:190147)
[error] 	at scala.concurrent.impl.Promise$Transformation.run(file:///tmp/shipreq.sbt/webapp-base-test/js/target/scala-2.13/webapp-base-test-test-fastopt.js:393398)
```

Does that help?

--- a/test-common/src/main/scala/org/scalajs/testing/common/Serializer.scala
+++ b/test-common/src/main/scala/org/scalajs/testing/common/Serializer.scala
@@ -93,8 +93,56 @@ private[testing] object Serializer {
   }
 
   implicit object StringSerializer extends Serializer[String] {
-    def serialize(x: String, out: SerializeState): Unit = out.out.writeUTF(x)
-    def deserialize(in: DeserializeState): String = in.in.readUTF()
+    def serialize(x: String, out: SerializeState): Unit = {
+      // Modified version of writeUTF to support strings longer than Short.MaxValue (#3667)
+      out.out.writeInt(x.length)
+
+      import out.out.write
+
+      for (i <- 0 until x.length()) {
+        val c = x.charAt(i)
+        if (c <= 0x7f && c >= 0x01) {
+          write(c)
+        } else if (c < 0x0800) {
+          write((c >> 6) | 0xc0)
+          write((c & 0x3f) | 0x80)
+        } else {
+          write((c >> 12) | 0xe0)
+          write(((c >> 6) & 0x3f) | 0x80)
+          write((c & 0x3f) | 0x80)
+        }
+      }
+    }
+
+    def deserialize(in: DeserializeState): String = {
+      import in.in.readByte
+
+      val chars = Array.fill(in.in.readInt()) {
+        val a = readByte()
+
+        if ((a & 0x80) == 0x00) { // 0xxxxxxx
+          a.toChar
+        } else if ((a & 0xE0) == 0xC0) { // 110xxxxx
+          val b = readByte()
+
+          require((b & 0xC0) == 0x80) // 10xxxxxx
+
+          (((a & 0x1F) << 6) | (b & 0x3F)).toChar
+        } else if ((a & 0xF0) == 0xE0) { // 1110xxxx
+          val b = readByte()
+          val c = readByte()
+
+          require((b & 0xC0) == 0x80) // 10xxxxxx
+          require((c & 0xC0) == 0x80) // 10xxxxxx
+
+          (((a & 0x0F) << 12) | ((b & 0x3F) << 6) | (c & 0x3F)).toChar
+        } else {
+          throw new IllegalArgumentException(s"bad byte: $a")
+        }
+      }
+
+      new String(chars)
+    }
   }
 
   implicit object UnitSerializer extends Serializer[Unit] {

--- a/test-common/src/test/scala/org/scalajs/testing/common/SerializerTest.scala
+++ b/test-common/src/test/scala/org/scalajs/testing/common/SerializerTest.scala
@@ -37,4 +37,12 @@ class SerializerTest {
     assertNull(deserialized.getFileName)
     assertEquals("MyClass.myMethod(Unknown Source)", deserialized.toString)
   }
+
+  // #3667
+  @Test
+  def serializeLargeString: Unit = {
+    val x = new String(Array.tabulate(Short.MaxValue + 1)(_.toChar))
+    val y = roundTrip(x)
+    assertEquals(x, y)
+  }
 }

